# 项目：整理Netflix电影演员评分数据

## 分析目标

此数据分析的目的是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，各演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

本实战项目的目的在于练习整理数据，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了截止至2022年7月美国地区可观看的所有Netflix电视剧及电影数据。数据集包含两个数据表：`titles.csv`和`credits.csv`。

`titles.csv`包含电影及电视剧相关信息，包括影视作品ID、标题、类型、描述、流派、IMDB（一个国外的在线评分网站）评分，等等。`credits.csv`包含超过7万名出现在Netflix影视作品的导演及演员信息，包括名字、影视作品ID、人物名、演职员类型（导演/演员）等。

`titles.csv`每列的含义如下：
- id：影视作品ID。
- title：影视作品标题。
- show_type：作品类型，电视节目或电影。
- description：简短描述。
- release_year：发布年份。
- age_certification：适龄认证。
- runtime：每集电视剧或电影的长度。
- genres：流派类型列表。
- production_countries：出品国家列表。
- seasons：如果是电视剧，则是季数。
- imdb_id：IMDB的ID。
- imdb_score：IMDB的评分。
- imdb_votes：IMDB的投票数。
- tmdb_popularity：TMDB的流行度。
- tmdb_score：TMDB的评分。

`credits.csv`每列的含义如下：
- person_ID：演职员ID。
- id：参与的影视作品ID。
- name：姓名。
- character_name：角色姓名。
- role：演职员类型，演员或导演。

## 读取数据

导入数据分析所需要的库，并通过Pandas的`read_csv`函数，将原始数据文件"titles.csv"里的数据内容，解析为DataFrame并赋值给变量`original_titles`。将原始数据文件"credits.csv"里的数据内容，解析为DataFrame并赋值给变量`original_credits`。

In [1]:
import pandas as pd

In [2]:
original_titles = pd.read_csv("titles.csv")
original_credits = pd.read_csv("credits.csv")

In [3]:
original_titles.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
4  tm120801                      The Dirty Dozen  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
4  12 American military prisoners in World War II...          1967   

  age_certification  runtime                                       genres  \
0             TV-MA       51                            ['documentation']   
1                 R      114                           ['drama', 'crime']   
2                 R      109  ['drama', 'action', 'thriller', 'european']   
3                PG       91              ['fantasy', 'action', 'comedy']   
4               NaN      150                            ['war', 'action']   

  production_countries  seasons    imdb_id  imdb_score  imdb_votes  \
0               ['US']      1.0        NaN         NaN         NaN   
1               ['US']      NaN  tt0075314         8.2    808582.0   
2               ['US']      NaN  tt0068473         7.7    107673.0   
3               ['GB']      NaN  tt0071853         8.2    534486.0   
4         ['GB', 'US']      NaN  tt0061578         7.7     72662.0   

   tmdb_popularity  tmdb_score  
0            0.600         NaN  
1           40.965       8.179  
2           10.010       7.300  
3           15.461       7.811  
4           20.398       7.600

In [4]:
original_credits.head()

person_id       id             name                character   role
0       3748  tm84618   Robert De Niro            Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster            Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                      Tom  ACTOR
3       3739  tm84618    Harvey Keitel  Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                    Betsy  ACTOR

## 评估和清洗数据

在这一部分中，我们将对在上一部分建立的original_titles及original_creditsDataFrame所包含的数据进行评估和清理。

主要从两个方面进行：结构和内容，即整齐度和干净度。

数据的结构性问题指不符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”这三个标准；数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

为了区分开经过清理的数据和原始的数据，我们创建新的变量cleaned_titles，让它为original_titles复制出的副本，以及创建新的变量cleaned_credits，让它为original_credits复制出的变量。我们之后的清理步骤都将被运用在cleaned_titles和cleaned_credits上。

In [5]:
cleaned_titles = original_titles.copy()
cleaned_credits = original_credits.copy()

### 数据整齐度

In [6]:
cleaned_titles.sample(10)

id                                    title   type  \
3215   tm433462                   We Summon the Darkness  MOVIE   
1131    ts42069                                The Crown   SHOW   
2784   tm408781                         Teefa in Trouble  MOVIE   
243     ts11313                               DEATH NOTE   SHOW   
5165  tm1045741                             Blue Miracle  MOVIE   
289     tm79895                                Halloween  MOVIE   
4582  tm1080109                             Quam's Money  MOVIE   
513     tm54147                                  Dostana  MOVIE   
2230   tm313988  The Death and Life of Marsha P. Johnson  MOVIE   
3344   ts222455                      The Devil Next Door   SHOW   

                                            description  release_year  \
3215  Three best friends attending a heavy-metal sho...          2020   
1131  The gripping, decades-spanning inside story of...          2016   
2784  Teefa goes to Poland to get Anya to Pakistan t...          2018   
243   Light Yagami is an ace student with great pros...          2006   
5165  To save their cash-strapped orphanage, a guard...          2021   
289   After being committed for 17 years, Michael My...          2007   
4582  In this sequel to "New Money," Quam, a securit...          2020   
513   Two straight guys who pretend to be a gay coup...          2008   
2230  Describing herself as a 'street queen,' Johnso...          2017   
3344  A Cleveland grandfather is brought to trial in...          2019   

     age_certification  runtime  \
3215                 R       91   
1131             TV-MA       56   
2784             PG-13      155   
243              TV-14       24   
5165               NaN       95   
289                  R      109   
4582               NaN      114   
513              PG-13      142   
2230               NaN      105   
3344             TV-MA       46   

                                                 genres production_countries  \
3215          ['thriller', 'horror', 'comedy', 'music']               ['US']   
1131                   ['drama', 'history', 'european']         ['GB', 'US']   
2784  ['drama', 'comedy', 'action', 'crime', 'romance']               ['PK']   
243   ['animation', 'thriller', 'scifi', 'drama', 'f...               ['JP']   
5165                      ['family', 'action', 'drama']         ['MX', 'US']   
289                                          ['horror']               ['US']   
4582                                          ['crime']               ['NG']   
513                      ['drama', 'romance', 'comedy']         ['IN', 'US']   
2230                                  ['documentation']               ['US']   
3344              ['crime', 'documentation', 'history']         ['US', 'IL']   

      seasons     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
3215      NaN   tt8058874         5.2     10483.0            8.183       5.500  
1131      5.0   tt4786824         8.7    194243.0           75.521       8.200  
2784      NaN   tt8043306         7.0      3659.0            3.176       7.600  
243       1.0   tt0877057         9.0    309445.0          120.545       8.664  
5165      NaN   tt7084386         6.6      6778.0           17.953       7.600  
289       NaN   tt0373883         6.0    119606.0           38.491       6.200  
4582      NaN         NaN         NaN         NaN            1.910       4.000  
513       NaN   tt1185420         6.5     15063.0            8.613       6.200  
2230      NaN   tt5233558         7.2      2964.0            6.792       7.300  
3344      1.0  tt11165002         7.6     13370.0            6.571       7.200

发现cleaned_titles表中的genres和production_countries这两个变量都有多个值，应该进行拆分

先提取任意一个genres变量的值进行观察。

In [9]:
cleaned_titles["genres"][1]

"['drama', 'crime']"

虽然genres表示形式是列表，但其实际类型并非字符串列表，而是字符串，无法直接用value_counts统计各个值出现的次数。 我们可以使用Python内置的eval函数，它可以把字符串转换成表达式，所以可以帮我们把表示列表的字符串转换成列表本身。

In [8]:
cleaned_titles["genres"] = cleaned_titles["genres"].apply(lambda x: eval(x))
cleaned_titles["genres"] [1]

['drama', 'crime']

== **转换为列表后，就能用DataFrame的explode方法，把那个列的列表值拆分成单独的行。**==

In [9]:
cleaned_titles = cleaned_titles.explode("genres")
cleaned_titles.head(5)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation               ['US']      1.0   
1                 R      114          drama               ['US']      NaN   
1                 R      114          crime               ['US']      NaN   
2                 R      109          drama               ['US']      NaN   
2                 R      109         action               ['US']      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

接下来，针对production_countries列也是一样的流程。

每个观察值的production_countries值并不表示单个流派，而是一系列流派。先提取任意一个production_countries变量的值进行观察。

In [10]:
cleaned_titles["production_countries"][2]

2    ['US']
2    ['US']
2    ['US']
2    ['US']
Name: production_countries, dtype: object

可以看到，production_countries也是一样的问题，虽然表示形式是列表，但其实际类型并非字符串列表，而是字符串，难以进行拆分。 我们可以再次利用eval函数进行类型转换，并检查转换后确实是列表类型。

In [11]:
cleaned_titles["production_countries"] = cleaned_titles["production_countries"].apply(lambda x :eval(x))
cleaned_titles.head(5)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                 [US]      1.0   
1                 R      114          drama                 [US]      NaN   
1                 R      114          crime                 [US]      NaN   
2                 R      109          drama                 [US]      NaN   
2                 R      109         action                 [US]      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300

确认类型转换完毕后，还是用explode方法，把列表值拆分成单独的行。

In [12]:
cleaned_titles = cleaned_titles.explode(["production_countries"])
cleaned_titles.head(10)

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
1   tm84618                          Taxi Driver  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
2  tm154986                          Deliverance  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   
3  tm127384      Monty Python and the Holy Grail  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
2  Intent on seeing the Cahulawassee River before...          1972   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   
3  King Arthur, accompanied by his squire, recrui...          1975   

  age_certification  runtime         genres production_countries  seasons  \
0             TV-MA       51  documentation                   US      1.0   
1                 R      114          drama                   US      NaN   
1                 R      114          crime                   US      NaN   
2                 R      109          drama                   US      NaN   
2                 R      109         action                   US      NaN   
2                 R      109       thriller                   US      NaN   
2                 R      109       european                   US      NaN   
3                PG       91        fantasy                   GB      NaN   
3                PG       91         action                   GB      NaN   
3                PG       91         comedy                   GB      NaN   

     imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0        NaN         NaN         NaN            0.600         NaN  
1  tt0075314         8.2    808582.0           40.965       8.179  
1  tt0075314         8.2    808582.0           40.965       8.179  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
2  tt0068473         7.7    107673.0           10.010       7.300  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811  
3  tt0071853         8.2    534486.0           15.461       7.811

在处理完cleaned_titles的结构性问题后，查看cleaned_credits

In [13]:
cleaned_credits.head(10)

person_id       id             name                      character   role
0       3748  tm84618   Robert De Niro                  Travis Bickle  ACTOR
1      14658  tm84618     Jodie Foster                  Iris Steensma  ACTOR
2       7064  tm84618    Albert Brooks                            Tom  ACTOR
3       3739  tm84618    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR
4      48933  tm84618  Cybill Shepherd                          Betsy  ACTOR
5      32267  tm84618      Peter Boyle                         Wizard  ACTOR
6     519612  tm84618   Leonard Harris      Senator Charles Palantine  ACTOR
7      29068  tm84618   Diahnne Abbott                Concession Girl  ACTOR
8     519613  tm84618      Gino Ardito             Policeman at Rally  ACTOR
9       3308  tm84618  Martin Scorsese  Passenger Watching Silhouette  ACTOR

从头部的10行数据来看，cleaned_credits数据符合“每个变量为一列，每个观察值为一行，每种类型的观察单位为一个表格”，因此不存在结构性问题。

## 数据干净度
接下来通过info，对数据内容进行大致了解。

In [14]:
cleaned_titles.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17818 entries, 0 to 5849
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    17818 non-null  object 
 1   title                 17817 non-null  object 
 2   type                  17818 non-null  object 
 3   description           17790 non-null  object 
 4   release_year          17818 non-null  int64  
 5   age_certification     10889 non-null  object 
 6   runtime               17818 non-null  int64  
 7   genres                17755 non-null  object 
 8   production_countries  17439 non-null  object 
 9   seasons               6224 non-null   float64
 10  imdb_id               17116 non-null  object 
 11  imdb_score            16976 non-null  float64
 12  imdb_votes            16945 non-null  float64
 13  tmdb_popularity       17663 non-null  float64
 14  tmdb_score            17241 non-null  float64
dtypes: float64(5), int64(2), 

从输出结果来看，cleaned_titles数据共有17818条观察值，title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量均存在缺失值，将在后续进行评估和清理。

此外，release_year表示年份，数据类型不应为数字，应为日期，所以需要进行数据格式转换。

In [15]:
cleaned_titles["release_year"] = pd.to_datetime(cleaned_titles["release_year"],format = "%Y")
cleaned_titles["release_year"] 

0      1945-01-01
1      1976-01-01
1      1976-01-01
2      1972-01-01
2      1972-01-01
          ...    
5847   2021-01-01
5848   2021-01-01
5849   2021-01-01
5849   2021-01-01
5849   2021-01-01
Name: release_year, Length: 17818, dtype: datetime64[ns]

format='%Y' 中的 %Y 是一个日期时间格式化代码，它表示将日期时间解析为四位数的年份。在日期时间字符串中，如果包含四位数的年份，可以使用 %Y 来表示。例如，如果日期时间字符串是 "2022-01-01"，那么 %Y 将解析为 "2022"。

In [16]:
cleaned_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77801 entries, 0 to 77800
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   person_id  77801 non-null  int64 
 1   id         77801 non-null  object
 2   name       77801 non-null  object
 3   character  68029 non-null  object
 4   role       77801 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.0+ MB


从输出结果来看，cleaned_credits数据共有77801条观察值，其中character变量存在缺失值，将在后续进行评估和清理。

此外，person_id表示演职员ID，数据类型不应为数字，应为字符串，所以需要进行数据格式转换。

In [17]:
cleaned_credits["person_id"] = cleaned_credits["person_id"].astype("str")
cleaned_credits["person_id"]

0           3748
1          14658
2           7064
3           3739
4          48933
          ...   
77796     736339
77797     399499
77798     373198
77799     378132
77800    1950416
Name: person_id, Length: 77801, dtype: object

### 处理缺失数据

在cleaned_titles中，title、description、age_certification、genres、production_countries、seasons、imdb_id、imdb_score、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量存在缺失值。

由于影视作品的标题、描述、适龄认证、发行国家、电视剧季数、IMDB的ID、TMDB的流行度、TMDB的评分，并不影响我们挖掘各个流派中的高IMDB评分作品演员，所以可以保留title、description、age_certification、production_countries、seasons、imdb_id、tmdb_popularity、tmdb_score、imdb_votes、tmdb_popularity、tmdb_score变量值存在空缺的观察值。

但imdb_score和genres，即IMDB评分和流派，和我们后续要做的分析息息相关。

先提取出imdb_score缺失观察值进行查看。

In [18]:
cleaned_titles[cleaned_titles["imdb_score"].isnull()]

id                                      title   type  \
0      ts300399        Five Came Back: The Reference Films   SHOW   
75     tm132164                       Bill Hicks: Sane Man  MOVIE   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
145    ts251477                            My First Errand   SHOW   
...         ...                                        ...    ...   
5810  tm1225897                                 Social Man  MOVIE   
5833   ts307884                                 HQ Barbers   SHOW   
5840  tm1216735                            Sun of the Soil  MOVIE   
5844  tm1074617              Bling Empire - The Afterparty  MOVIE   
5848  tm1035612  Dad Stop Embarrassing Me - The Afterparty  MOVIE   

                                            description release_year  \
0     This collection includes 12 World War II-era p...   1945-01-01   
75    Sane Man was filmed before Bill recorded ‘Dang...   1989-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
145   “Hajimete no Otsukai” (First Errand) is a Japa...   1991-01-01   
...                                                 ...          ...   
5810  Two competitive social media Influencers go he...   2021-01-01   
5833  When a family run barber shop in the heart of ...   2021-01-01   
5840  In 14th-century Mali, an ambitious young royal...   2022-01-01   
5844  The stars of "Bling Empire" discuss the show's...   2021-01-01   
5848  Jamie Foxx, David Alan Grier and more from the...   2021-01-01   

     age_certification  runtime         genres production_countries  seasons  \
0                TV-MA       51  documentation                   US      1.0   
75                   R       80         comedy                   US      NaN   
145               TV-G       18  documentation                   JP     12.0   
145               TV-G       18         family                   JP     12.0   
145               TV-G       18        reality                   JP     12.0   
...                ...      ...            ...                  ...      ...   
5810               NaN       96          drama                  NaN      NaN   
5833             TV-14       24         comedy                   NG      1.0   
5840               NaN       26            NaN                  NaN      NaN   
5844               NaN       35            NaN                   US      NaN   
5848             PG-13       37            NaN                   US      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0            NaN         NaN         NaN            0.600         NaN  
75           NaN         NaN         NaN            3.377         7.5  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
145          NaN         NaN         NaN            7.730         7.8  
...          ...         ...         ...              ...         ...  
5810  tt20198164         NaN         NaN              NaN         NaN  
5833         NaN         NaN         NaN            0.840         NaN  
5840         NaN         NaN         NaN            1.179         7.0  
5844         NaN         NaN         NaN              NaN         NaN  
5848         NaN         NaN         NaN            1.296        10.0  

[842 rows x 15 columns]

由于缺失分析所需的核心数据imdb_score，我们将把这些观察值删除，并查看删除后该列空缺值个数和：

In [19]:
cleaned_titles = cleaned_titles.dropna(subset = ["imdb_score"])
cleaned_titles["imdb_score"].isnull().sum()

0

然后提取出genres缺失观察值进行查看。

In [20]:
cleaned_titles.query("genres.isnull()")

id                                              title   type  \
1813    ts77824  My Next Guest Needs No Introduction With David...   SHOW   
1939   ts215037                              Minecraft: Story Mode   SHOW   
2386    ts74805                   A Little Help with Carol Burnett   SHOW   
2658   ts265844                                           #ABtalks   SHOW   
4274  tm1172010                                  The Lockdown Plan  MOVIE   
4648  tm1113921                                           In Vitro  MOVIE   

                                            description release_year  \
1813  TV legend David Letterman teams up with fascin...   2018-01-01   
1939  MInecraft: Story Mode is an interactive, anima...   2018-01-01   
2386  In this unscripted series starring comedy lege...   2018-01-01   
2658  #ABtalks is a YouTube interview show hosted by...   2018-01-01   
4274                                                NaN   2020-01-01   
4648  'In Vitro' is an otherworldly rumination on me...   2019-01-01   

     age_certification  runtime genres production_countries  seasons  \
1813             TV-MA       50    NaN                   US      4.0   
1939             TV-PG       52    NaN                   US      1.0   
2386              TV-G       24    NaN                   US      1.0   
2658             TV-PG       68    NaN                  NaN      1.0   
4274               NaN       49    NaN                  NaN      NaN   
4648               NaN       27    NaN                  NaN      NaN   

         imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
1813   tt7829834         7.8      5581.0            8.217         7.6  
1939  tt10498322         5.6       347.0              NaN         NaN  
2386   tt7204366         6.3       237.0            1.621         6.2  
2658  tt12635254         9.6         7.0              NaN         NaN  
4274  tt13079112         6.5         NaN              NaN         NaN  
4648  tt10545994         7.7         NaN              NaN         NaN

由于缺失分析所需的核心数据genres，我们将把这些观察值删除，并查看删除后该列空缺值个数和：

In [21]:
cleaned_titles = cleaned_titles.dropna(subset = ["genres"])
cleaned_titles["genres"].isnull().sum()

0

接下来评估cleaned_credits的缺失数据，其中只有character变量存在缺失值。

角色名并不影响我们挖掘各个流派中的高IMDB评分作品演员，并且此变量缺失也有可能因为演职员类别是导演，没有对应角色，因此可以保留character变量值存在空缺的观察值。

### 处理重复数据

根据数据变量的含义以及内容来看，cleaned_titles里不应该存在每个变量值都相同的观察值，因此查看是否存在重复值。

In [22]:
cleaned_titles.duplicated().sum()

0

输出结果为0，说明不存在重复值。

接下来查看cleaned_credits数据表是否存在重复值。

In [23]:
cleaned_credits.duplicated().sum()

0

输出结果为0，说明不存在重复值。

### 处理不一致数据

针对cleaned_titles，不一致数据可能存在于genres和character变量中，我们将查看是否存在多个不同值指代同一流派，以及多个不同值指代同一国家的情况。

In [24]:
cleaned_titles['genres'].value_counts()

genres
drama            3357
comedy           2419
thriller         1446
action           1339
romance          1080
crime            1066
documentation     981
family            769
animation         732
fantasy           727
european          679
scifi             647
horror            438
history           336
music             266
reality           226
war               221
sport             188
western            53
Name: count, dtype: int64

从上面看出，genres列里并不存在不一致数据，各个值都在指代不同的流派。但是里面还存在空字符串表示的流派，并非有效数据，因此可以进行删除。

删除后，查看cleaned_titles里是否还存在genres为空字符串的行：

In [25]:
cleaned_titles = cleaned_titles.query('genres != ""')
cleaned_titles.query('genres == ""')['genres'].sum()

0

接下来，针对production_countries列也是一样的流程，利用value_counts方法，得到production_countries的列表里面各个值的出现次数。

In [26]:
cleaned_titles['production_countries].value_counts()

SyntaxError: EOL while scanning string literal (427808933.py, line 1)

由于value_counts执行结果中有太多值，Pandas只会默认显示开头和结尾的一些值。要完整展示结果，可以把display.max_rows设置为None，即取消展示行数上限。

但因为我们只是在当前调用value_counts时才需要看完整结果，所以可以结合option_context，只更改临时上限。

使用 with pd.option_context() 在特定代码块中临时更改 Pandas 的显示选项，以确保只在需要时显示所有行和列：

python
Copy code
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(df)  # 在此代码块中打印 DataFrame

In [ ]:
with pd.option_context('display.max_rows', None):
    print(cleaned_titles['production_countries'].value_counts())

从以上输出结果来看，出品国家都用两位的国家代码来表示，除了里面存在一个的Lebanon值。

Lebanon的国家代码是LB，出现了39次，说明此处数据不一致。LB和Lebanon都在表示同一国家，需要进行统一。

把cleaned_titles里，production_countries的"LB"和"Lebanon"统一为LB，并检查替换后是否还存在"LB"：

In [ ]:
cleaned_titles['production_countries'].replace('Lebanon','LB',inplace = True)
(cleaned_titles['production_countries']=='LB').sum()

另外，里面还存在空字符串表示的国家代码，并非有效数据。但由于出品国家并非分析所需的关键信息，所以可以保留出品国家为空的观察值。

针对original_credits，不一致数据可能存在于role中，我们将查看是否存在多个不同值指代同一演职员类型的情况。

In [ ]:
original_credits['role'].value_counts()

从以上输出结果来看，role只有两种可能的值，ACTOR或DIRECTOR，不存在不一致数据。我们可以把这列的类型转换为Category，好处是比字符串类型更节约内存空间，也能表明说值的类型有限。

In [ ]:
cleaned_credits["role"] = cleaned_credits["role"].astype("category")
cleaned_credits["role"]

处理无效或错误数据

可以通过DataFrame的describe方法，对数值统计信息进行快速了解

In [ ]:
cleaned_titles.describe()

## 整理数据

In [ ]:
cleaned_titles

In [ ]:
cleaned_credits

对数据的整理，与分析方向紧密相关。此次数据分析目标是，整理不同流派影视作品，比如喜剧片、动作片、科幻片中，演员出演作品的平均IMDB评分，从而挖掘出各个流派中的高评分作品演员。

那为了能同时获得流派与演员数据，我们需要把cleaned_titles和cleaned_credits，通过id作为键进行连接，因为两个数据表中id都是影视作品ID。

In [29]:
credits_with_titles = pd.merge(cleaned_credits,cleaned_titles, on="id",how='inner')
credits_with_titles

person_id         id                name      character      role  \
0           3748    tm84618      Robert De Niro  Travis Bickle     ACTOR   
1           3748    tm84618      Robert De Niro  Travis Bickle     ACTOR   
2          14658    tm84618        Jodie Foster  Iris Steensma     ACTOR   
3          14658    tm84618        Jodie Foster  Iris Steensma     ACTOR   
4           7064    tm84618       Albert Brooks            Tom     ACTOR   
...          ...        ...                 ...            ...       ...   
276104    736339  tm1059008    Adelaida Buscato      María Paz     ACTOR   
276105    399499  tm1059008  Luz Stella Luengas   Karen Bayona     ACTOR   
276106    373198  tm1059008         Inés Prieto          Fanny     ACTOR   
276107    378132  tm1059008        Isabel Gaona         Cacica     ACTOR   
276108   1950416  tm1059008      Julian Gaviria            NaN  DIRECTOR   

              title   type                                        description  \
0       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
1       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
2       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
3       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
4       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
...             ...    ...                                                ...   
276104      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276105      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276106      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276107      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276108      Lokillo  MOVIE  A controversial TV host and comedian who has b...   

       release_year age_certification  runtime  genres production_countries  \
0        1976-01-01                 R      114   drama                   US   
1        1976-01-01                 R      114   crime                   US   
2        1976-01-01                 R      114   drama                   US   
3        1976-01-01                 R      114   crime                   US   
4        1976-01-01                 R      114   drama                   US   
...             ...               ...      ...     ...                  ...   
276104   2021-01-01               NaN       90  comedy                   CO   
276105   2021-01-01               NaN       90  comedy                   CO   
276106   2021-01-01               NaN       90  comedy                   CO   
276107   2021-01-01               NaN       90  comedy                   CO   
276108   2021-01-01               NaN       90  comedy                   CO   

        seasons     imdb_id  imdb_score  imdb_votes  tmdb_popularity  \
0           NaN   tt0075314         8.2    808582.0           40.965   
1           NaN   tt0075314         8.2    808582.0           40.965   
2           NaN   tt0075314         8.2    808582.0           40.965   
3           NaN   tt0075314         8.2    808582.0           40.965   
4           NaN   tt0075314         8.2    808582.0           40.965   
...         ...         ...         ...         ...              ...   
276104      NaN  tt14585902         3.8        68.0           26.005   
276105      NaN  tt14585902         3.8        68.0           26.005   
276106      NaN  tt14585902         3.8        68.0           26.005   
276107      NaN  tt14585902         3.8        68.0           26.005   
276108      NaN  tt14585902         3.8        68.0           26.005   

        tmdb_score  
0            8.179  
1            8.179  
2            8.179  
3            8.179  
4            8.179  
...            ...  
276104       6.300  
276105       6.300  
276106       6.300  
276107       6.300  
276108       6.300  

[276109 rows x 19 columns]

连接后，我们就能知道各个演职员参与过的影视作品的具体信息。


In [31]:
actor_with_titles = credits_with_titles.query('role=="ACTOR"')
actor_with_titles

person_id         id                name      character   role  \
0           3748    tm84618      Robert De Niro  Travis Bickle  ACTOR   
1           3748    tm84618      Robert De Niro  Travis Bickle  ACTOR   
2          14658    tm84618        Jodie Foster  Iris Steensma  ACTOR   
3          14658    tm84618        Jodie Foster  Iris Steensma  ACTOR   
4           7064    tm84618       Albert Brooks            Tom  ACTOR   
...          ...        ...                 ...            ...    ...   
276103    368473  tm1059008        Aída Morales        Maritza  ACTOR   
276104    736339  tm1059008    Adelaida Buscato      María Paz  ACTOR   
276105    399499  tm1059008  Luz Stella Luengas   Karen Bayona  ACTOR   
276106    373198  tm1059008         Inés Prieto          Fanny  ACTOR   
276107    378132  tm1059008        Isabel Gaona         Cacica  ACTOR   

              title   type                                        description  \
0       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
1       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
2       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
3       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
4       Taxi Driver  MOVIE  A mentally unstable Vietnam War veteran works ...   
...             ...    ...                                                ...   
276103      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276104      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276105      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276106      Lokillo  MOVIE  A controversial TV host and comedian who has b...   
276107      Lokillo  MOVIE  A controversial TV host and comedian who has b...   

       release_year age_certification  runtime  genres production_countries  \
0        1976-01-01                 R      114   drama                   US   
1        1976-01-01                 R      114   crime                   US   
2        1976-01-01                 R      114   drama                   US   
3        1976-01-01                 R      114   crime                   US   
4        1976-01-01                 R      114   drama                   US   
...             ...               ...      ...     ...                  ...   
276103   2021-01-01               NaN       90  comedy                   CO   
276104   2021-01-01               NaN       90  comedy                   CO   
276105   2021-01-01               NaN       90  comedy                   CO   
276106   2021-01-01               NaN       90  comedy                   CO   
276107   2021-01-01               NaN       90  comedy                   CO   

        seasons     imdb_id  imdb_score  imdb_votes  tmdb_popularity  \
0           NaN   tt0075314         8.2    808582.0           40.965   
1           NaN   tt0075314         8.2    808582.0           40.965   
2           NaN   tt0075314         8.2    808582.0           40.965   
3           NaN   tt0075314         8.2    808582.0           40.965   
4           NaN   tt0075314         8.2    808582.0           40.965   
...         ...         ...         ...         ...              ...   
276103      NaN  tt14585902         3.8        68.0           26.005   
276104      NaN  tt14585902         3.8        68.0           26.005   
276105      NaN  tt14585902         3.8        68.0           26.005   
276106      NaN  tt14585902         3.8        68.0           26.005   
276107      NaN  tt14585902         3.8        68.0           26.005   

        tmdb_score  
0            8.179  
1            8.179  
2            8.179  
3            8.179  
4            8.179  
...            ...  
276103       6.300  
276104       6.300  
276105       6.300  
276106       6.300  
276107       6.300  

[262614 rows x 19 columns]

由于我们只对挖掘演员的参演作品口碑感兴趣，导演不在我们的分析范围内，因此根据role，筛选出类型为ACTOR的观察值，供后续分析。

接下来我们要进行分组操作，根据genre和person_id进行分组

In [34]:
groupby_genres_and_person_id = actor_with_titles.groupby(["genres","person_id"])["imdb_score"].mean()
groupby_genres_and_person_id 

genres   person_id
action   1000         6.866667
         100007       7.000000
         100013       6.400000
         100019       6.500000
         100020       6.500000
                        ...   
western  993735       6.500000
         998673       7.300000
         998674       7.300000
         998675       7.300000
         99940        4.000000
Name: imdb_score, Length: 168881, dtype: float64

转换为df

In [35]:
groupby_genres_and_person_id_df = groupby_genres_and_person_id.reset_index()

In [36]:
groupby_genres_and_person_id_df 

genres person_id  imdb_score
0        action      1000    6.866667
1        action    100007    7.000000
2        action    100013    6.400000
3        action    100019    6.500000
4        action    100020    6.500000
...         ...       ...         ...
168876  western    993735    6.500000
168877  western    998673    7.300000
168878  western    998674    7.300000
168879  western    998675    7.300000
168880  western     99940    4.000000

[168881 rows x 3 columns]

根据genres分组，选取出imd最大值

In [37]:
genres_max_scores = groupby_genres_and_person_id_df.groupby(["genres"])["imdb_score"].max()

In [38]:
genres_max_scores

genres
action           9.3
animation        9.3
comedy           9.2
crime            9.5
documentation    9.1
drama            9.5
european         8.9
family           9.3
fantasy          9.3
history          9.1
horror           9.0
music            8.8
reality          8.9
romance          9.2
scifi            9.3
sport            9.1
thriller         9.5
war              8.8
western          8.9
Name: imdb_score, dtype: float64

用此表继续找到原先的人物

In [39]:
genres_max_score_with_person_id = pd.merge(genres_max_scores,groupby_genres_and_person_id_df, on=["genres","imdb_score"])
genres_max_score_with_person_id 

genres  imdb_score person_id
0     action         9.3     12790
1     action         9.3      1303
2     action         9.3     21033
3     action         9.3    336830
4     action         9.3     86591
..       ...         ...       ...
131      war         8.8    826547
132  western         8.9     22311
133  western         8.9     28166
134  western         8.9     28180
135  western         8.9     93017

[136 rows x 3 columns]

继续将名字匹配上，先把cleaned_credits修理一下，只用person_id和name

In [43]:
actor_id_names = cleaned_credits[["person_id","name"]].drop_duplicates()
actor_id_names

person_id                name
0          3748      Robert De Niro
1         14658        Jodie Foster
2          7064       Albert Brooks
3          3739       Harvey Keitel
4         48933     Cybill Shepherd
...         ...                 ...
77796    736339    Adelaida Buscato
77797    399499  Luz Stella Luengas
77798    373198         Inés Prieto
77799    378132        Isabel Gaona
77800   1950416      Julian Gaviria

[54589 rows x 2 columns]

In [44]:
genres_maxscore_actor_name = pd.merge(genres_max_score_with_person_id ,actor_id_names, on=["person_id"])

In [45]:
genres_maxscore_actor_name

genres  imdb_score person_id                name
0       action         9.3     12790         Olivia Hack
1        scifi         9.3     12790         Olivia Hack
2       action         9.3      1303       Jessie Flower
3    animation         9.3      1303       Jessie Flower
4       family         9.3      1303       Jessie Flower
..         ...         ...       ...                 ...
131        war         8.8    826547         Yuto Uemura
132    western         8.9     22311     Koichi Yamadera
133    western         8.9     28166  Megumi Hayashibara
134    western         8.9     28180      Unsho Ishizuka
135    western         8.9     93017            Aoi Tada

[136 rows x 4 columns]

按照geners排序

In [51]:
final_data = genres_maxscore_actor_name.sort_values("genres").reset_index().drop("index",axis=1)

In [52]:
final_data

genres  imdb_score person_id                name
0     action         9.3     12790         Olivia Hack
1     action         9.3    336830     André Sogliuzzo
2     action         9.3     21033          Zach Tyler
3     action         9.3     86591       Cricket Leigh
4     action         9.3      1303       Jessie Flower
..       ...         ...       ...                 ...
131      war         8.8    826547         Yuto Uemura
132  western         8.9     28166  Megumi Hayashibara
133  western         8.9     28180      Unsho Ishizuka
134  western         8.9     22311     Koichi Yamadera
135  western         8.9     93017            Aoi Tada

[136 rows x 4 columns]